File: 42-trapping-rain-water/42-trapping-rain-water.py
# result = 0
class Solution:
    
    def trap(self, height: List[int]) -> int:
        start = 0
        end = len(height)-1
        result = 0
        leftTarget = 0
        rightTarget = 0
        while start< end:
            if height[start] <= height[end]:
                if leftTarget < height[start]:
                    leftTarget = height[start]
                else:
                    result +=(leftTarget - height[start])
                start += 1
            else:
                if rightTarget < height[end]:
                    rightTarget = height[end]
                else:
                    result += (rightTarget - height[end])
                end -= 1
                
        return result

File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        Stack<Integer> stack = new Stack<Integer>();
	int begin = 0, result = 0;
	while( begin < height.length && height[begin] == 0 )
		begin++;
	    for(int i = begin; i < height.length; i++) {
		    while( !stack.isEmpty() && height[i] > height[stack.peek()] ) {
			    int valley = height[stack.pop()];
    			if( !stack.isEmpty() ){
                    result += (Math.min(height[i], height[stack.peek()]) - valley) * (i - stack.peek() - 1);
                }
		    }
		stack.push(i);
	    }
	return result;
    }
}